internal/pkg/runtime/engine/singularity/prepare_linux.go

```diff
@@ -1009,22 +1010,31 @@ func (e *EngineOperations) setSessionLayer(img *image.Image) error {
 		return nil
 	}
 
+	// Check for implicit user namespace, e.g when we run %test in a fakeroot build
+	// https://github.com/hpcng/singularity/issues/5315
+	userNS, _ := namespaces.IsInsideUserNamespace(os.Getpid())
+
 	// NEED FIX: on ubuntu until 4.15 kernel it was possible to mount overlay
 	// with the current workflow, since 4.18 we get an operation not permitted
 	for _, ns := range e.EngineConfig.OciConfig.Linux.Namespaces {
 		if ns.Type == specs.UserNamespace {
-			if !e.EngineConfig.File.EnableUnderlay {
-				sylog.Debugf("Not attempting to use underlay with user namespace: disabled by configuration ('enable underlay = no')")
-				return nil
-			}
-			if !writableImage {
-				sylog.Debugf("Using underlay layer: user namespace requested")
-				e.EngineConfig.SetSessionLayer(singularityConfig.UnderlayLayer)
-				return nil
-			}
-			sylog.Debugf("Not attempting to use overlay or underlay: writable flag requested")
+			userNS = true
+			break
```

Comment:
If `userNS` is `true`, we don't need to iterate in this loop. To kill two birds with one stone, how about:

```
	for _, ns := range e.EngineConfig.OciConfig.Linux.Namespaces; !userNS {
		if ns.Type == specs.UserNamespace {
			userNS = true
		}
	}
```

Reply:
Adding `; !userNS` isn't valid on a `for xxx := range xxx`

Will add the extra if. Didn't as the loop is going to be over one item (mount namespace) in most cases, so not much perf benefit, but I agree it shows the intent.